### ISSUE TITLE
`notmuch new` triggers "Mailbox was externally modified" warning in neomutt, even when no new mail

### ISSUE THREAD

**Issue by brianbuccola** [neomutt/neomutt#1793]
I've just migrated from mutt to neomutt + notmuch, and there's one problem that I can't solve.

When I run `notmuch new`, neomutt *always* issues the following warning

```
Mailbox was externally modified.  Flags may be wrong.
```

even when `notmuch new` doesn't actually (seem to) modify the mailbox:

```bash
$ notmuch new
No new mail.
```

Notably, the same warning does *not* occur in regular mutt, so I guess it's a warning about the *virtual* mailbox that I'm viewing in neomutt.

It's particularly annoying because it also issues an audible beep + visual bell, which I use to notify me of new mail.

Is this a bug, or have I configured something wrong? It may be related to #170, but I can't tell.

(FWIW, I'm only using notmuch to create a single virtual mailbox/folder that combines two regular mailboxes/folders. Nothing too fancy.)

**Comment by Austin-Ray** [comment]
Can you attach the output of `neomutt -v`?

**Comment by brianbuccola** [comment]
```
NeoMutt 20180716
Copyright (C) 1996-2016 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.2.3-arch1-1-ARCH (x86_64)
ncurses: ncurses 6.1.20180127 (compiled with 6.1.20180127)
libidn2: 2.2.0 (compiled with 2.1.1)
hcache backends: kyotocabinet, gdbm, lmdb

Compiler:
Using built-in specs.
COLLECT_GCC=cc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/8.2.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --enable-multilib --disable-werror --enable-checking=release --enable-default-pie --enable-default-ssp --enable-cet=auto
Thread model: posix
gcc version 8.2.1 20181127 (GCC) 

Configure options: --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --gpgme --enable-lua --notmuch --gss --ssl --sasl --with-ui=ncurses --with-idn2=/usr --idn=0 --idn2=1 --lmdb --kyotocabinet --gdbm

Compilation CFLAGS: -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include  -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens +getaddrinfo 
  -gnutls +gpgme +gss +hcache -homespool +idn -locales_hack +lua +meta 
  -mixmaster +nls +notmuch +openssl +pgp +sasl +smime +start_color 
  +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

**Comment by Austin-Ray** [comment]
Thanks for the output, @brianbuccola.

The `Mailbox was externally modified.  Flags may be wrong.` message is intended when running `notmuch new` if the underlying mailbox changes. It's to warn the user that what they're seeing may not be 100% accurate. All mailbox backends do this.

The message appearing when `notmuch new` does not modified anything is a bug. I'll look into it further since doesn't appear 100% of the time for me.

**Comment by Austin-Ray** [comment]
@flatcap The root cause looks to be pretty complicated. I will need your assistance debugging it. 

Here's the story:

1. `nm_mbox_check()` returns `MUTT_REOPENED`, which triggers the `Mailbox was externally modified.  Flags may be wrong.` warning message. This is because `occult` is set to true at: https://github.com/neomutt/neomutt/blob/417c0eae4f377db0d61fc4713ff386cc08302223/notmuch/mutt_notmuch.c#L2304-L2312
2. In `nm_mbox_check()`, we loop over each email and set `active` to `true` except in the case when the email is `NULL`. https://github.com/neomutt/neomutt/blob/417c0eae4f377db0d61fc4713ff386cc08302223/notmuch/mutt_notmuch.c#L2265-L2271 This moves the cause to `get_mutt_email()` since we're getting `e == NULL`
3. Inside `get_mutt_email()`, there are three obvious places that return `NULL`; however, it's the final `return` that's `NULL`. I added a debug print statement to verify. This means `mutt_hash_find()` is missing emails (29 out of 51250 in my case.)

Any suggestions for debugging the missing hash elements?

**Comment by flatcap** [comment]
> Any suggestions for debugging the missing hash elements?

Hmm... nothing inspiring.

Your notmuch query returns 51250 results.  Does the underlying mailbox have that many emails?
Are there any duplicate Message-Ids?  That would mess up the hash table.
Are there any messages with no Message-Id? (either invalid emails, or badly parsed by NeoMutt)

**Comment by Austin-Ray** [comment]
> Your notmuch query returns 51250 results. Does the underlying mailbox have that many emails?

Yes, `notmuch count` confirmed results.

> Are there any duplicate Message-Ids?

Yes, there are duplicate message-ids; however, `notmuch` reports I have tons of duplicates. My maildir has 31616 emails with one duplicate, 2904 with two duplicates, etc. `notmuch` handles de-duplication on queries. The issue only occurs for 29 emails consistently.

> Are there any messages with no Message-Id?

It's possible, but I've confirmed the 29 emails have matching, valid message-ids.

This is quite vexing.

**Comment by flatcap** [comment]
I'd add an `if` that matches the 29 ghost emails and breakpoint the dummy payload,
(conditional breakpoints can be really slow)
then trace through what happens to the email.

**Comment by Austin-Ray** [comment]
@flatcap: I've been investigating this further without much progress. I've noticed the 29 missing emails aren't missing at all. I can find them in the index alongside their duplicates! 

Every time I trigger the "Mailbox was externally modified. Flags may be wrong." message, a new duplicate appears in the email thread.

Something must be occurring at a low-level than the `notmuch`-backend.

**Comment by Austin-Ray** [comment]
@flatcap: I've found the culprit. The entries are in the hash table, but with a different key. The key is the message-id, and this is set in `mutt_notmuch.c`'s `init_email`:

https://github.com/neomutt/neomutt/blob/55392f6c7c8ca4b40e16ce6fed1944f20d322c19/notmuch/mutt_notmuch.c#L779-L783

`init_email()` takes in an email and will update the message-id if it is `NULL`; however, it does not handle the case where the message-id differs from what `notmuch` reports. An edge case exists in this one spot. 

Changing line 781 to 
```C
if (!e->env->message_id  || (e->env->message_id != nm2mutt_message_id(id)))
```
will update the message-id to what notmuch reports and stops the `Mailbox was externally modified. Flags may be wrong.` message from appearing when running `notmuch new --full-scan --verbose`.

The reason for the provided email being incorrect is unknown, but it appears in 0.05% of case, which I believe qualifies as an edge-case.

**Comment by ceuk** [comment]
Hi @flatcap @Austin-Ray  - sorry to necrobump this issue but I'm getting exactly the same issue - is the cause likely to be a bug or a config issue on my part?

**Comment by brianbuccola** [comment]
I still have the same issue. I had switched back to normal mutt before @Austin-Ray and @flatcap fixed this in #1797, so I can't confirm whether that commit actually solved this specific problem (and I haven't used neomutt since then), but in any case, I just switched back to neomutt + notmuch to test, and I can confirm that I still see the same behavior.

**Comment by brianbuccola** [comment]
Hi, even though this issue is now closed, I wanted to give a quick update: on a whim (mainly because I totally forgot about this issue), I updated my mailsync script to both sync my mail *and* then run `notmuch new` (every few minutes), just as I described directly above 3 years ago (starting with "Ideally..."). Now, 3 years later, I still get the "Mailbox was externally modified" message even when there's no new mail, but I **no longer** get a beep/bell alert (unless there really is new mail). So, I call that a win. Thanks so much to everyone who helped solve this issue!
